File: latest/week2/workshop6_m9/W06_M09_demo.R
# ============================================================
# Workshop 6 — Module 9: Case Study — e-Car Temporal Analysis
# Integrated R Script — UNYT Tirana
# ============================================================
library(tidyverse); library(patchwork); dir.create("output", showWarnings = FALSE)

# ── 1. LOAD AND PREPARE ─────────────────────────────────
ec <- read_csv("ecar.csv")
# Clean column names (dataset has double-space in "Car  Type")
names(ec) <- str_replace_all(names(ec), "\\s+", " ") |> str_trim()

ec <- ec |>
  mutate(
    ApproveDate = mdy(`Approve Date`),
    Year = year(ApproveDate),
    Month = month(ApproveDate),
    Quarter = quarter(ApproveDate),
    YearMonth = floor_date(ApproveDate, "month"),
    Spread = Rate - `Cost of Funds`,
    Approved = Outcome == 1) |>
  filter(Year >= 2002, Year <= 2012)

cat("Dataset:", nrow(ec), "loans, Years:", min(ec$Year), "-", max(ec$Year), "\n")
cat("Columns:", paste(names(ec), collapse = ", "), "\n")

# ── 2. YEARLY RATE + SPREAD WITH 2008 ANNOTATION ────────
yearly <- ec |>
  group_by(Year) |>
  summarise(
    mean_rate = mean(Rate, na.rm = TRUE),
    mean_spread = mean(Spread, na.rm = TRUE),
    mean_cof = mean(`Cost of Funds`, na.rm = TRUE),
    volume = n(),
    approval_rate = mean(Approved, na.rm = TRUE),
    .groups = "drop")

p_rate_spread <- ggplot(yearly) +
  geom_line(aes(x = Year, y = mean_rate, color = "Offered Rate"),
    linewidth = 1.5) +
  geom_point(aes(x = Year, y = mean_rate), color = "#1565C0", size = 2.5) +
  geom_line(aes(x = Year, y = mean_spread, color = "Spread (Rate - CoF)"),
    linewidth = 1.5) +
  geom_point(aes(x = Year, y = mean_spread), color = "#E53935", size = 2.5) +
  geom_line(aes(x = Year, y = mean_cof, color = "Cost of Funds"),
    linewidth = 1, linetype = "dashed") +
  scale_color_manual(values = c(
    "Offered Rate" = "#1565C0",
    "Spread (Rate - CoF)" = "#E53935",
    "Cost of Funds" = "#2E7D32")) +
  # 2008 crisis annotation
  geom_vline(xintercept = 2008, linetype = "dashed", color = "#888", linewidth = 0.6) +
  annotate("rect", xmin = 2007.5, xmax = 2009.5, ymin = -Inf, ymax = Inf,
    fill = "#E53935", alpha = 0.05) +
  annotate("text", x = 2008, y = max(yearly$mean_rate) * 0.95,
    label = "2008\nFinancial\nCrisis", size = 2.5, color = "#888",
    fontface = "bold") +
  scale_x_continuous(breaks = 2002:2012) +
  theme_minimal() +
  labs(title = "e-Car Loans: Rate, Cost of Funds, and Spread (2002–2012)",
       subtitle = "Rate dropped post-2008 (Fed cuts); Spread compressed then recovered",
       x = "Year", y = "Percentage Points", color = NULL)
ggsave("output/rate_spread_cof.png", p_rate_spread, width = 10, height = 5, dpi = 300)

# ── 3. VOLUME AND APPROVAL RATE BAR CHARTS ──────────────
p_volume <- ggplot(yearly, aes(x = Year, y = volume / 1000)) +
  geom_col(fill = "#2E7D32", width = 0.6, alpha = 0.7) +
  geom_text(aes(label = paste0(round(volume/1000, 1), "k")),
    vjust = -0.5, size = 2.5, fontface = "bold") +
  geom_vline(xintercept = 2008, linetype = "dashed", color = "#888") +
  scale_x_continuous(breaks = 2002:2012) +
  theme_minimal() +
  labs(title = "(a) Loan Volume (thousands per year)",
       x = "Year", y = "Loans (000s)")

p_approval <- ggplot(yearly, aes(x = Year, y = approval_rate * 100)) +
  geom_line(color = "#7B1FA2", linewidth = 1.5) +
  geom_point(color = "#7B1FA2", size = 3) +
  geom_text(aes(label = paste0(round(approval_rate * 100, 1), "%")),
    vjust = -1, size = 2.5, color = "#7B1FA2") +
  geom_vline(xintercept = 2008, linetype = "dashed", color = "#888") +
  scale_x_continuous(breaks = 2002:2012) +
  theme_minimal() +
  labs(title = "(b) Approval Rate (%)",
       x = "Year", y = "Approval Rate (%)")

ggsave("output/volume_approval.png", p_volume | p_approval,
  width = 12, height = 5, dpi = 300)

# ── 4. QUARTERLY RATE WITH CONFIDENCE RIBBON ─────────────
quarterly <- ec |>
  group_by(Year, Quarter) |>
  summarise(
    mean_rate = mean(Rate, na.rm = TRUE),
    se_rate = sd(Rate, na.rm = TRUE) / sqrt(n()),
    n = n(), .groups = "drop") |>
  mutate(
    date = as.Date(paste0(Year, "-", Quarter * 3 - 2, "-01")),
    lower = mean_rate - 1.96 * se_rate,
    upper = mean_rate + 1.96 * se_rate)

p_quarterly <- ggplot(quarterly, aes(x = date)) +
  geom_ribbon(aes(ymin = lower, ymax = upper),
    fill = "#1565C0", alpha = 0.15) +
  geom_line(aes(y = mean_rate), color = "#1565C0", linewidth = 0.8) +
  geom_point(aes(y = mean_rate), color = "#1565C0", size = 1) +
  # Lehman collapse
  geom_vline(xintercept = as.Date("2008-09-15"),
    linetype = "dashed", color = "#E53935", linewidth = 0.6) +
  annotate("text", x = as.Date("2008-12-01"),
    y = max(quarterly$upper) * 0.95,
    label = "Lehman\ncollapse", size = 2.5, color = "#E53935",
    fontface = "bold", hjust = 0) +
  scale_x_date(date_breaks = "1 year", date_labels = "%Y") +
  theme_minimal() +
  labs(title = "Quarterly Mean Rate ± 95% CI",
       subtitle = "CI widens during crisis = more pricing volatility",
       x = NULL, y = "Rate (%)")
ggsave("output/quarterly_ci.png", p_quarterly, width = 10, height = 5, dpi = 300)

# ── 5. RATE BY TIER OVER TIME (small multiples) ─────────
tier_yearly <- ec |>
  group_by(Year, Tier) |>
  summarise(mean_rate = mean(Rate, na.rm = TRUE), .groups = "drop")

p_tier_multi <- ggplot(tier_yearly, aes(x = Year, y = mean_rate)) +
  geom_line(color = "#1565C0", linewidth = 1) +
  geom_point(color = "#1565C0", size = 1.5) +
  geom_vline(xintercept = 2008, linetype = "dashed", color = "#888", linewidth = 0.3) +
  facet_wrap(~paste0("Tier ", Tier), ncol = 5) +
  scale_x_continuous(breaks = c(2002, 2005, 2008, 2012)) +
  theme_minimal(base_size = 8) +
  labs(title = "Rate by Credit Tier Over Time (2002–2012)",
       subtitle = "All tiers dropped post-2008; Tier 1 dropped furthest (competitive pressure)",
       x = "Year", y = "Mean Rate (%)")
ggsave("output/tier_multiples.png", p_tier_multi, width = 14, height = 4, dpi = 300)

# Grey+accent: highlight Tier 1
p_tier_accent <- ggplot(tier_yearly, aes(x = Year, y = mean_rate, group = factor(Tier))) +
  geom_line(color = "#DDDDDD", linewidth = 0.7) +
  geom_line(data = tier_yearly |> filter(Tier == 1),
    color = "#E53935", linewidth = 2) +
  geom_text(data = tier_yearly |>
    group_by(Tier) |> filter(Year == max(Year)),
    aes(label = paste0("T", Tier),
        color = Tier == 1),
    hjust = -0.3, size = 2.5, fontface = "bold", show.legend = FALSE) +
  scale_color_manual(values = c("TRUE" = "#E53935", "FALSE" = "#AAAAAA")) +
  geom_vline(xintercept = 2008, linetype = "dashed", color = "#888") +
  scale_x_continuous(breaks = 2002:2012, limits = c(2002, 2013)) +
  theme_minimal() +
  labs(title = "Rate by Tier: Grey+Accent on Tier 1 (best credit)",
       subtitle = "Tier 1 dropped most aggressively — banks competed for safest borrowers",
       x = "Year", y = "Mean Rate (%)")
ggsave("output/tier_accent.png", p_tier_accent, width = 10, height = 5, dpi = 300)

# ── 6. SPREAD DISTRIBUTION: PRE vs POST 2008 ────────────
ec_period <- ec |>
  mutate(period = ifelse(Year < 2008, "Pre-2008 (2002–2007)", "Post-2008 (2008–2012)"))

p_spread_dist <- ggplot(ec_period, aes(x = Spread, fill = period)) +
  geom_histogram(bins = 50, alpha = 0.5, position = "identity", color = "white") +
  scale_fill_manual(values = c(
    "Pre-2008 (2002–2007)" = "#1565C0",
    "Post-2008 (2008–2012)" = "#E53935")) +
  geom_vline(data = ec_period |> group_by(period) |>
    summarise(med = median(Spread)),
    aes(xintercept = med, color = period),
    linetype = "dashed", linewidth = 1, show.legend = FALSE) +
  scale_color_manual(values = c(
    "Pre-2008 (2002–2007)" = "#1565C0",
    "Post-2008 (2008–2012)" = "#E53935")) +
  theme_minimal() +
  labs(title = "Spread Distribution: Pre vs Post 2008",
       subtitle = "Post-2008 spread compressed (dashed = median)",
       x = "Spread (pp)", y = "Count", fill = "Period")
ggsave("output/spread_pre_post.png", p_spread_dist, width = 9, height = 5, dpi = 300)

# ── 7. MONTHLY RATE DECOMPOSITION (STL) ─────────────────
# Monthly average rate
monthly_rate <- ec |>
  group_by(YearMonth) |>
  summarise(mean_rate = mean(Rate, na.rm = TRUE), .groups = "drop") |>
  arrange(YearMonth)

# STL requires ts object
ts_rate <- ts(monthly_rate$mean_rate, frequency = 12,
  start = c(year(min(monthly_rate$YearMonth)),
    month(min(monthly_rate$YearMonth))))
stl_fit <- stl(ts_rate, s.window = "periodic")

# Extract components
components <- tibble(
  date = monthly_rate$YearMonth,
  observed = as.numeric(ts_rate),
  trend = as.numeric(stl_fit$time.series[, "trend"]),
  seasonal = as.numeric(stl_fit$time.series[, "seasonal"]),
  remainder = as.numeric(stl_fit$time.series[, "remainder"]))

p_decomp <- components |>
  pivot_longer(-date, names_to = "component", values_to = "value") |>
  mutate(component = factor(component,
    levels = c("observed", "trend", "seasonal", "remainder"))) |>
  ggplot(aes(x = date, y = value)) +
  geom_line(aes(color = component), linewidth = 0.5, show.legend = FALSE) +
  scale_color_manual(values = c(
    observed = "#333", trend = "#1565C0",
    seasonal = "#2E7D32", remainder = "#E53935")) +
  facet_wrap(~component, ncol = 1, scales = "free_y") +
  geom_vline(xintercept = as.Date("2008-09-15"),
    linetype = "dashed", color = "#888", linewidth = 0.3) +
  theme_minimal(base_size = 8) +
  labs(title = "STL Decomposition: Monthly Mean Rate (2002–2012)",
       subtitle = "Trend captures the 2008 crisis; seasonal shows within-year pattern",
       x = NULL, y = NULL)
ggsave("output/rate_stl.png", p_decomp, width = 10, height = 8, dpi = 300)

# ── 8. SIX-PANEL DASHBOARD ──────────────────────────────
dashboard <- (p_rate_spread) /
  (p_volume | p_approval) /
  (p_quarterly | p_tier_accent) +
  plot_annotation(
    title = "e-Car Temporal Case Study: Six-Panel Dashboard (2002–2012)",
    subtitle = "Rate dynamics + Volume + Approval + Quarterly CI + Tier stratification",
    tag_levels = "a")
ggsave("output/ecar_dashboard.png", dashboard, width = 14, height = 14, dpi = 300)

# ── 9. KEY FINDINGS ──────────────────────────────────────
cat("\n── KEY FINDINGS ──\n")
cat("1. Rate peak:", round(max(yearly$mean_rate), 2), "% in",
  yearly$Year[which.max(yearly$mean_rate)], "\n")
cat("2. Rate trough:", round(min(yearly$mean_rate), 2), "% in",
  yearly$Year[which.min(yearly$mean_rate)], "\n")
cat("3. Spread pre-2008 median:", round(median(ec$Spread[ec$Year < 2008]), 2), "pp\n")
cat("4. Spread post-2008 median:", round(median(ec$Spread[ec$Year >= 2008]), 2), "pp\n")
cat("5. Volume peak:", format(max(yearly$volume), big.mark = ","), "in",
  yearly$Year[which.max(yearly$volume)], "\n")
cat("6. Approval rate range:", round(min(yearly$approval_rate)*100, 1), "–",
  round(max(yearly$approval_rate)*100, 1), "%\n")

cat("\n── All W06-M09 R outputs saved ──\n")

File: latest/week2/workshop6_m9/W06_M09_demo.py
"""W06-M09: Case Study — e-Car Temporal Analysis — Python — UNYT"""
import numpy as np, pandas as pd, matplotlib.pyplot as plt, os
from matplotlib.dates import DateFormatter
from matplotlib.gridspec import GridSpec
os.makedirs("output", exist_ok=True)

# ── 1. LOAD AND PREPARE ─────────────────────────────────
ec = pd.read_csv("ecar.csv")
ec.columns = [c.strip().replace("  ", " ") for c in ec.columns]
ec["ApproveDate"] = pd.to_datetime(ec["Approve Date"], format="%m/%d/%Y", errors="coerce")
ec["Year"] = ec["ApproveDate"].dt.year
ec["Month"] = ec["ApproveDate"].dt.month
ec["Quarter"] = ec["ApproveDate"].dt.quarter
ec["YearMonth"] = ec["ApproveDate"].dt.to_period("M").dt.to_timestamp()
ec["Spread"] = ec["Rate"] - ec["Cost of Funds"]
ec["Approved"] = ec["Outcome"] == 1
ec = ec.query("2002 <= Year <= 2012")

print(f"Dataset: {len(ec)} loans, {ec['Year'].min()}–{ec['Year'].max()}")

# ── 2. YEARLY RATE + SPREAD + CoF ───────────────────────
yr = ec.groupby("Year").agg(
    rate=("Rate", "mean"), spread=("Spread", "mean"),
    cof=("Cost of Funds", "mean"), volume=("Rate", "size"),
    approval=("Approved", "mean")).reset_index()

fig, ax = plt.subplots(figsize=(10, 5))
ax.plot(yr.Year, yr.rate, "-o", color="#1565C0", lw=1.5, markersize=5, label="Offered Rate")
ax.plot(yr.Year, yr.spread, "-s", color="#E53935", lw=1.5, markersize=5, label="Spread")
ax.plot(yr.Year, yr.cof, "--^", color="#2E7D32", lw=1, markersize=4, label="Cost of Funds")
# 2008 annotation
ax.axvspan(2007.5, 2009.5, alpha=0.05, color="#E53935")
ax.axvline(x=2008, ls="--", color="#888", lw=0.6)
ax.text(2008, yr.rate.max() * 0.95, "2008\nCrisis", fontsize=8,
    fontweight="bold", color="#888", ha="center")
ax.legend(fontsize=8)
ax.set_title("e-Car: Rate, Cost of Funds, and Spread (2002–2012)\n"
    "Rate dropped post-2008 (Fed cuts); Spread compressed then recovered",
    fontweight="bold")
ax.set_xlabel("Year"); ax.set_ylabel("Percentage Points")
ax.set_xticks(range(2002, 2013))
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
plt.tight_layout(); plt.savefig("output/rate_spread_cof.png", dpi=300); plt.close()

# ── 3. VOLUME + APPROVAL ────────────────────────────────
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(12, 5))

ax1.bar(yr.Year, yr.volume / 1000, color="#2E7D32", width=0.6, alpha=0.7)
for _, row in yr.iterrows():
    ax1.text(row.Year, row.volume/1000 + 0.3, f"{row.volume/1000:.1f}k",
        ha="center", fontsize=6, fontweight="bold")
ax1.axvline(x=2008, ls="--", color="#888")
ax1.set_title("(a) Loan Volume (thousands/year)", fontweight="bold")
ax1.set_ylabel("Loans (000s)"); ax1.set_xticks(range(2002, 2013))
ax1.tick_params(labelsize=7)

ax2.plot(yr.Year, yr.approval * 100, "-o", color="#7B1FA2", lw=1.5, markersize=5)
for _, row in yr.iterrows():
    ax2.text(row.Year, row.approval*100 + 0.8, f"{row.approval*100:.1f}%",
        ha="center", fontsize=6, color="#7B1FA2", fontweight="bold")
ax2.axvline(x=2008, ls="--", color="#888")
ax2.set_title("(b) Approval Rate (%)", fontweight="bold")
ax2.set_ylabel("Approval %"); ax2.set_xticks(range(2002, 2013))
ax2.tick_params(labelsize=7)

for ax in [ax1, ax2]:
    ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
plt.tight_layout(); plt.savefig("output/volume_approval.png", dpi=300); plt.close()

# ── 4. QUARTERLY RATE WITH CI RIBBON ─────────────────────
qr = ec.groupby(["Year", "Quarter"]).agg(
    rate=("Rate", "mean"), se=("Rate", "sem"), n=("Rate", "size")).reset_index()
qr["date"] = pd.to_datetime(
    qr.apply(lambda r: f"{int(r.Year)}-{int(r.Quarter)*3-2:02d}-01", axis=1))
qr["lower"] = qr.rate - 1.96 * qr.se
qr["upper"] = qr.rate + 1.96 * qr.se

fig, ax = plt.subplots(figsize=(10, 5))
ax.fill_between(qr.date, qr.lower, qr.upper, alpha=0.15, color="#1565C0")
ax.plot(qr.date, qr.rate, color="#1565C0", lw=0.8)
ax.axvline(x=pd.Timestamp("2008-09-15"), ls="--", color="#E53935", lw=0.8)
ax.annotate("Lehman\ncollapse", xy=(pd.Timestamp("2008-09-15"), qr.rate.max() * 0.95),
    xytext=(pd.Timestamp("2009-03-01"), qr.rate.max() * 0.90),
    fontsize=8, fontweight="bold", color="#E53935",
    arrowprops=dict(arrowstyle="->", color="#E53935", lw=0.8))
ax.set_title("Quarterly Mean Rate ± 95% CI\n"
    "CI widens during crisis = more pricing volatility", fontweight="bold")
ax.set_ylabel("Rate (%)"); ax.xaxis.set_major_formatter(DateFormatter("%Y"))
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
plt.tight_layout(); plt.savefig("output/quarterly_ci.png", dpi=300); plt.close()

# ── 5. RATE BY TIER — SMALL MULTIPLES ───────────────────
tiers = sorted(ec["Tier"].unique())
fig, axes = plt.subplots(1, len(tiers), figsize=(14, 4), sharey=True)
for ax, tier in zip(axes, tiers):
    ty = ec.query("Tier == @tier").groupby("Year")["Rate"].mean()
    ax.plot(ty.index, ty.values, "-o", color="#1565C0", lw=1, markersize=3)
    ax.axvline(x=2008, ls="--", color="#888", lw=0.3)
    ax.set_title(f"Tier {tier}", fontsize=9, fontweight="bold")
    ax.set_xticks([2002, 2005, 2008, 2012]); ax.tick_params(labelsize=5)
    ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
axes[0].set_ylabel("Mean Rate (%)", fontsize=8)
fig.suptitle("Rate by Credit Tier (2002–2012): All tiers dropped post-2008",
    fontsize=12, fontweight="bold")
plt.tight_layout(); plt.savefig("output/tier_multiples.png", dpi=300); plt.close()

# Grey+accent on Tier 1
fig, ax = plt.subplots(figsize=(10, 5))
tier_yr = ec.groupby(["Year", "Tier"])["Rate"].mean().reset_index()
for tier in tiers:
    sub = tier_yr.query("Tier == @tier")
    ax.plot(sub.Year, sub.Rate, color="#DDDDDD", lw=0.7)
sub1 = tier_yr.query("Tier == 1")
ax.plot(sub1.Year, sub1.Rate, color="#E53935", lw=2.5, label="Tier 1 (best credit)")
for tier in tiers:
    sub = tier_yr.query("Tier == @tier")
    color = "#E53935" if tier == 1 else "#AAAAAA"
    ax.text(sub.Year.values[-1] + 0.15, sub.Rate.values[-1],
        f"T{tier}", fontsize=7, fontweight="bold", color=color, va="center")
ax.axvline(x=2008, ls="--", color="#888")
ax.set_title("Rate by Tier: Grey+Accent on Tier 1\n"
    "Tier 1 dropped most (banks competed for safest borrowers)", fontweight="bold")
ax.set_xlabel("Year"); ax.set_ylabel("Mean Rate (%)")
ax.set_xticks(range(2002, 2013)); ax.set_xlim(2001.5, 2013.5)
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
plt.tight_layout(); plt.savefig("output/tier_accent.png", dpi=300); plt.close()

# ── 6. SPREAD: PRE vs POST 2008 ─────────────────────────
fig, ax = plt.subplots(figsize=(9, 5))
pre = ec.query("Year < 2008")["Spread"]
post = ec.query("Year >= 2008")["Spread"]
ax.hist(pre, bins=50, alpha=0.5, color="#1565C0",
    label=f"Pre-2008 (med={pre.median():.2f})")
ax.hist(post, bins=50, alpha=0.5, color="#E53935",
    label=f"Post-2008 (med={post.median():.2f})")
ax.axvline(x=pre.median(), ls="--", color="#1565C0", lw=1.5)
ax.axvline(x=post.median(), ls="--", color="#E53935", lw=1.5)
ax.legend(fontsize=9)
ax.set_title("Spread Distribution: Pre vs Post 2008\n"
    "Post-2008 spread compressed (dashed = median)", fontweight="bold")
ax.set_xlabel("Spread (pp)"); ax.set_ylabel("Count")
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
plt.tight_layout(); plt.savefig("output/spread_pre_post.png", dpi=300); plt.close()

# ── 7. STL DECOMPOSITION OF MONTHLY RATE ────────────────
try:
    from statsmodels.tsa.seasonal import STL
    monthly_rate = ec.groupby("YearMonth")["Rate"].mean().sort_index()
    stl = STL(monthly_rate, period=12, robust=True)
    res = stl.fit()

    fig, axes = plt.subplots(4, 1, figsize=(10, 8), sharex=True)
    panels = [(monthly_rate, "Observed", "#333"),
              (res.trend, "Trend", "#1565C0"),
              (res.seasonal, "Seasonal", "#2E7D32"),
              (res.resid, "Residual", "#E53935")]
    for ax, (data, title, color) in zip(axes, panels):
        ax.plot(data.index, data.values, color=color, lw=0.6)
        ax.axvline(x=pd.Timestamp("2008-09-15"), ls="--", color="#888", lw=0.3)
        ax.set_title(title, fontsize=9, fontweight="bold", color=color, loc="left")
        ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
        ax.tick_params(labelsize=6)
    fig.suptitle("STL Decomposition: Monthly Mean Rate (2002–2012)", fontweight="bold")
    plt.tight_layout(); plt.savefig("output/rate_stl.png", dpi=300); plt.close()
    print("STL decomposition saved")
except ImportError:
    print("statsmodels not available — skipping STL")

# ── 8. KEY FINDINGS ──────────────────────────────────────
print(f"\n=== KEY FINDINGS ===")
print(f"1. Rate peak: {yr.rate.max():.2f}% in {int(yr.loc[yr.rate.idxmax(),'Year'])}")
print(f"2. Rate trough: {yr.rate.min():.2f}% in {int(yr.loc[yr.rate.idxmin(),'Year'])}")
print(f"3. Spread pre-2008 median: {pre.median():.2f}pp")
print(f"4. Spread post-2008 median: {post.median():.2f}pp")
print(f"5. Volume peak: {int(yr.volume.max()):,} in {int(yr.loc[yr.volume.idxmax(),'Year'])}")
print(f"6. Approval range: {yr.approval.min()*100:.1f}–{yr.approval.max()*100:.1f}%")

print("\nAll W06-M09 Python outputs saved")
